finality-aleph/src/sync/handler.rs

```diff
@@ -0,0 +1,486 @@
+    }
+}
+
+impl<I: PeerId, J: Justification, CS: ChainStatus<J>, V: Verifier<J>, F: Finalizer<J>>
+    Handler<I, J, CS, V, F>
+{
+    /// New handler with the provided chain interfaces.
+    pub fn new(
+        chain_status: CS,
+        verifier: V,
+        finalizer: F,
+        period: SessionPeriod,
+    ) -> Result<Self, Error<J, CS, V, F>> {
+        let top_finalized = chain_status
+            .top_finalized()
+            .map_err(Error::ChainStatus)?
+            .header()
+            .id();
+        let forest = Forest::new(top_finalized);
+        Ok(Handler {
+            chain_status,
+            verifier,
+            finalizer,
+            forest,
+            period,
+        })
+    }
+
+    fn large_justification_batch_from(
+        &self,
+        id: BlockIdFor<J>,
+    ) -> Result<NetworkData<J>, Error<J, CS, V, F>> {
+        use Error::*;
+        let mut result = Vec::with_capacity(MAX_SMALL_JUSTIFICATION_BATCH as usize);
+        let mut number = id.number() + 1;
+        let top_finalized_number = self
+            .chain_status
+            .top_finalized()
+            .map_err(ChainStatus)?
+            .header()
+            .id()
+            .number();
+        while result.len() < MAX_JUSTIFICATION_BATCH as usize && number <= top_finalized_number {
+            number = match self
+                .chain_status
+                .finalized_at(number)
+                .map_err(ChainStatus)?
+            {
+                Some(justification) => {
+                    result.push(justification.into_unverified());
+                    number + 1
+                }
+                // We might skip all blocks of a session if we are missing a justification, but
+                // this will happen only for sessions where we don't have all the justifications.
+                // The last block of a session was always guaranteed to contain a justification, so
+                // we only share that one. It can be missing only if the last block is above the
+                // top finalized, in which case this will break the loop.
+                None => last_block_of_session(
+                    session_id_from_block_num(number, self.period),
+                    self.period,
+                ),
+            }
+        }
+        Ok(NetworkData::Justifications(result, self.state()?))
+    }
+
+    fn small_justification_batch_from(
+        &self,
+        id: BlockIdFor<J>,
+    ) -> Result<NetworkData<J>, Error<J, CS, V, F>> {
+        let mut result = Vec::with_capacity(MAX_SMALL_JUSTIFICATION_BATCH as usize);
+        let mut number = id.number() + 1;
+        while result.len() < MAX_SMALL_JUSTIFICATION_BATCH as usize {
+            match self
+                .chain_status
+                .finalized_at(number)
+                .map_err(Error::ChainStatus)?
+            {
+                Some(justification) => result.push(justification.into_unverified()),
+                None => break,
```

Comment:
Is there a reason not to send last block of this session here if it exists? Seems to me like if we did that here, we could just make one function `justification_batch_for` which would take batch size as an argument.

Reply:
Hmm, for short fresh batches this means that we are above the top finalized block, as they have justifications for all the blocks. So this would never send the last block in the sense you are implying above. It is also a much simpler break condition than in the other method.

After looking at this for a while I agree that the methods could be unified into one (the more complicated one), with the behaviour remaining identical in non-bugged situations. I am not sure this is that helpful though – it seems to me that for readability having the usual situation be represented by a much simpler method, and only have weird edge cases (needed only for ancient sessions) in the more rarely used one is better. But I'm not very convinced either way – what do you two think? cc @lesniak43

Reply:
yes, in typical situation it would not send the justification, and if we used the logic of a more complicated method, it would just ended the loop, as last block of session would be over top finalized. 

I just feel, like there could be some strange state we did not expect and in this case our node would not send the justifications. That said, it would probably quickly fix itself, as soon as gap between two nodes is big enough.

So I guess we could leave two different functions here.